Firestore/core/src/firebase/firestore/util/executor.h

```diff
@@ -0,0 +1,128 @@
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_EXECUTOR_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_EXECUTOR_H_
+
+#include <chrono>  // NOLINT(build/c++11)
+#include <functional>
+#include <string>
+#include <utility>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+// A handle to an operation scheduled for future execution. The handle may
+// outlive the operation, but it *cannot* outlive the executor that created it.
+class DelayedOperation {
+ public:
+  DelayedOperation() {
+  }
+
+  // If the operation has not been run yet, cancels the operation. Otherwise,
+  // this function is a no-op.
+  void Cancel() {
+    cancel_func_();
+  }
+
+  // Internal use only.
+  explicit DelayedOperation(std::function<void()>&& cancel_func)
+      : cancel_func_{std::move(cancel_func)} {
+  }
+
+ private:
+  std::function<void()> cancel_func_;
+};
+
+namespace internal {
+
+// An interface to a platform-specific executor of asynchronous tasks
+// ("operations").
+//
+// Operations may be scheduled for immediate or delayed execution. Operations
+// scheduled for the exact same time are FIFO ordered.
+//
+// The operations are executed sequentially; only a single operation is executed
+// at any given time.
+//
+// Delayed operations may be canceled if they have not already been run.
+class Executor {
+ public:
+  using Tag = int;
+  using Operation = std::function<void()>;
+  using Milliseconds = std::chrono::milliseconds;
+
+  // Operations scheduled for future execution are tagged to allow retreiving
+  // the later. The tag is entirely opaque for the executor; in particular,
+  // uniqueness of tags is not enforced.
+  struct TaggedOperation {
+    TaggedOperation() {
+    }
+    TaggedOperation(const Tag tag, Operation&& operation)
+        : tag{tag}, operation{std::move(operation)} {
+    }
+    Tag tag = 0;
+    Operation operation;
+  };
+
+  virtual ~Executor() {
+  }
+
+  // Schedules the `operation` to be asynchronously executed as soon as
+  // possible. If called in quick succession, the operations will be
+  // FIFO-ordered.
+  virtual void Execute(Operation&& operation) = 0;
+  // Like `Execute`, but blocks until the `operation` finishes, consequently
+  // draining immediate operations from the executor.
+  virtual void ExecuteBlocking(Operation&& operation) = 0;
+  // Scheduled the given `operation` to be executed after `delay` milliseconds
+  // from now, and returns a handle that allows to cancel the operation
+  // (provided it hasn't been run already). The operation is tagged to allow
+  // retrieving it later.
+  //
+  // `delay` must be non-negative; use `Execute` to schedule operations for
+  // immediate execution.
+  virtual DelayedOperation ScheduleExecution(Milliseconds delay,
+                                             TaggedOperation&& operation) = 0;
+
+  // Checks for the caller whether it is being invoked by this executor.
+  virtual bool IsAsyncCall() const = 0;
```

Comment:
This is equivalent to isCurrentQueue before, right? This name doesn't really convey that it's an equivalent test. It seems like it might be trying to tell the difference between `Execute` vs `ExecuteBlocking` (but of course that's nonsensical).

Maybe `IsCurrentExecutor`?